### PR TITLE
Allow variable stacklevel for deprecation warnings.

### LINKED ISSUE #6137: Incorrect stacklevel for  deprecate_multichannel_kwarg  deprecation warnings.
## Description

deprecate_multichannel_kwarg and alike properly set the stacklevel to 2, 
But for functions with this decorator are also decorated with `@utils.channel_as_last_axis()`, which adds at  1 level between the caller and the final warning location.

This mean that 1) the warning messages are less useful, they all point to `skimage/_shared/utils.py:338`.
2) filter set for dependees  turn warnings into errors don't become error (e.g Napari)

## Way to reproduce
```python
# foo.py
from skimage.transform import pyramid_gaussian
pyramid_gaussian(None, multichannel=True)
```
```
$ python foo.py
/Users/bussonniermatthias/dev/scikit-image/skimage/_shared/utils.py:338: FutureWarning: `multichannel` is a deprecated argument name for `pyramid_gaussian`. It will be removed in version 1.0.Please use `channel_axis` instead.
  return func(*args, **kwargs)
```

## Version information

e37fc660aeb96d53db27c9e2f1a0a88920bc2b65 (dec 15th), orther info 

deprecate_multichannel_kwarg could gain and "extra_stacklevel" to properly place the warning where relevant.

### PR DESCRIPTION
**Pull request by Carreau**
Each decoration needs to be manually checked, I have ideas on how to
make that better, by inspecting the above frames and look whether there
is a sentinel asking to add 1 to the stack number, but not tonight.

closes #6137

Alternative would be to use something like:

```python
def extra_stack():
     import sys
     f = sys._getframe()
     fback = f.f_back
     count = 0
     while '__DEPRECATION_SKIP_STACK__' in fback.f_locals.keys():
         count +=1
         fback = fback.f_back
     return count

```

And mark corresponding decorators with ``__DEPRECATION_SKIP_STACK__ = 1``.
or 
```python
def extra_stack(module_name='skimage'):
     import sys
     f = sys._getframe()
     fback = f.f_back
     count = 0
     while module_name in fback.f_global['__name__']:
         count +=1
         fback = fback.f_back
     return count
```

Which would always compute stacklevels so that deprecation warnings are emitted outside of skimage, potentially with a global toggle for skimage own CI Suite.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
i like the explicit parameter approach.

Adding additional global variables is tricky to get by correct, and I feel like it won't scale well with additional support for other interpreters.

i don't have time to read the code now in full but i do think it is an interesting approach.

the other thing i would say, i often as a keyword only argument in some of my functions called: 
- `_stacklevel_increment`
to allow one to forward warnings an arbitrary amount up the call stack.

File: skimage/restoration/_denoise.py [hunk 1]
```diff
@@ -258,9 +258,9 @@ def denoise_bilateral(image, win_size=None, sigma_color=None, sigma_spatial=1,
 
 
 @utils.channel_as_last_axis()
-@utils.deprecate_multichannel_kwarg()
+@utils.deprecate_multichannel_kwarg(extra_stacklevel=1)
 @utils.deprecate_kwarg({'max_iter': 'max_num_iter'}, removed_version="1.0",
-                       deprecated_version="0.19")
+                       deprecated_version="0.19", extra_stacklevel=1)
```
Comment (rfezzani) [inline skimage/restoration/_denoise.py:263]:
```suggestion
                       deprecated_version="0.19")
```
This is causing the CI failures: `extra_stacklevel` is defined for `deprecate_kawarg`

**Comment by rfezzani** [comment]
@Carreau, as an alternative, please see #6183 that defines a decorator's base class for storing the callstack length of each decorated function to automatically compute the correct warning `stacklevel`.

File: skimage/_shared/utils.py [hunk 2]
```diff
@@ -208,7 +208,7 @@ def __init__(self, kwarg_mapping, deprecated_version, warning_msg=None,
                                 "for `{func_name}`. ")
             if removed_version is not None:
                 self.warning_msg += (f'It will be removed in '
-                                     f'version {removed_version}.')
+                                     f'version {removed_version}. ')
```
Comment (mkcor) [inline skimage/_shared/utils.py:211]: @rfezzani please cherry-pick this for #6183 :wink:

Comment (rfezzani) [reply skimage/_shared/utils.py:211]: Done, thank you :wink:

**Comment by grlee77** [comment]
closing this as we ended up going with the approach in #6183. Thanks for initiating this work, @Carreau

**Pull request closed without merging**